blackbird_python/blackbird/program.py

```diff
@@ -0,0 +1,290 @@
+            float: version number
+        """
+        return self._version
+
+    @property
+    def modes(self):
+        """A set of non-negative integers specifying the mode numbers the program manipulates.
+
+        Returns:
+            set[int]: mode numbers
+        """
+        return self._modes
+
+    @property
+    def target(self):
+        """Contains information regarding the target device of the quantum
+        program (i.e., the target device the Blackbird script is compiled for).
+
+        Important keys include:
+
+        * ``'name'`` (str): the name of the device the Blackbird script requests to be
+            run on. If no target is requested, the returned value will be ``None``.
+        * ``'options'`` (dict): a dictionary of keyword arguments for the target device
+
+        Returns:
+            dict[str->[str, dict]]: target information
+        """
+        return self._target
+
+    @property
+    def operations(self):
+        """List of operations to apply to the device, in temporal order.
+
+        Each operation is contained as a dictionary, with the following keys:
+
+        * ``'op'`` (str): the name of the operation
+        * ``'args'`` (list): a list of positional arguments for the operation
+        * ``'kwargs'`` (dict): a dictionary of keyword arguments for the operation
+        * ``'modes'`` (list[int]): modes the operation applies to
+
+        Note that, depending on the operation, both ``'args'`` and ``'kwargs'``
+        might be empty.
+
+        Returns:
+            list[dict]: operation information
+        """
+        return self._operations
+
+    def __len__(self):
+        """The length of the quantum program (i.e., the number of operations applied).
+
+        Returns:
+            int: program length
+        """
+        return len(self._operations)
+
+    def serialize(self):
+        """Serializes the blackbird program, returning a valid Blackbird script
+        as a string.
+
+        Returns:
+            str: the blackbird script representing the BlackbirdProgram object
+        """
+        # top level metadata
+        var_count = 0
+        array_insert = 3
+
+        script = ["name {}".format(self.name), "version {}".format(self.version)]
+
+        if self.target["name"] is not None:
+            array_insert += 1
+            options = ""
+
+            if self.target["options"]:
+                # if the target has options, compile them into
+                # the expected syntax
+                option_strings = [
+                    "{}={}".format(k, v)
+                    if not isinstance(v, str)
+                    else '{}="{}"'.format(k, v)
+                    for k, v in self.target["options"].items()
+                ]
+                options = " ({})".format(", ".join(option_strings))
+
+            # add target metadata
+            script.append("target {}{}".format(self.target["name"], options))
+
+        # line break
+        script.append("")
```

Comment:
not "\n"?

Reply:
Nope, each list element is assumed to be a new line.